src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,511 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
+                                //
+                                var A = left;
+                                var y = rightDispatch;
+                                var B = y.Left;
+                                var z = y.Right;
+                                return y.WithLeftAndRight(new RelationalDispatch(syntax, value, op, A, B), z);
+                            }
+                        }
+                        #endregion Rebalance the top of the tree if necessary
+
+                        // that should have been sufficient to balance the tree.
+                        Debug.Assert(Math.Abs(left.Height - right.Height) < 2);
+                        return new RelationalDispatch(syntax, value, op, left: left, right: right);
+                    }
+                }
+            }
+
+            private void LowerValueDispatchNode(
+                ValueDispatchNode n,
+                BoundDagTemp input)
+            {
+                var inputExpression = _tempAllocator.GetTemp(input);
+                LowerValueDispatchNodeCore(n, inputExpression);
+            }
+
+            private void LowerValueDispatchNodeCore(ValueDispatchNode n, BoundExpression input)
+            {
+                switch (n)
+                {
+                    case ValueDispatchNode.LeafDispatchNode leaf:
+                        _loweredDecisionDag.Add(_factory.Goto(leaf.Label));
+                        return;
+                    case ValueDispatchNode.SwitchDispatch eq:
+                        LowerSwitchDispatchNode(eq, input);
+                        return;
+                    case ValueDispatchNode.RelationalDispatch rel:
+                        LowerRelationalDispatchNode(rel, input);
+                        return;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(n);
+                }
+            }
+
+            private void LowerRelationalDispatchNode(ValueDispatchNode.RelationalDispatch rel, BoundExpression input)
+            {
+                var test = MakeRelationalTest(rel.Syntax, input, rel.Operator, rel.Value);
+                if (rel.WhenTrue is ValueDispatchNode.LeafDispatchNode whenTrue)
+                {
+                    LabelSymbol trueLabel = whenTrue.Label;
+                    _loweredDecisionDag.Add(_factory.ConditionalGoto(test, trueLabel, jumpIfTrue: true));
+                    LowerValueDispatchNodeCore(rel.WhenFalse, input);
+                }
+                else if (rel.WhenFalse is ValueDispatchNode.LeafDispatchNode whenFalse)
+                {
+                    LabelSymbol falseLabel = whenFalse.Label;
+                    _loweredDecisionDag.Add(_factory.ConditionalGoto(test, falseLabel, jumpIfTrue: false));
+                    LowerValueDispatchNodeCore(rel.WhenTrue, input);
+                }
+                else
+                {
+                    LabelSymbol falseLabel = _factory.GenerateLabel("sw");
```

Comment:
nit: We've used more explicit label names in other places. Consider doing it here too. Same for "gt" label below in this file.